crates/driver/src/boundary/liquidity/uniswap/v3.rs

```diff
@@ -0,0 +1,132 @@
+    },
+    anyhow::Result,
+    bigdecimal::ToPrimitive,
+    contracts::{GPv2Settlement, UniswapV3SwapRouter},
+    shared::{
+        current_block::BlockRetrieving,
+        http_solver::model::TokenAmount,
+        interaction::Interaction,
+        sources::uniswap_v3::pool_fetching::UniswapV3PoolFetcher,
+    },
+    solver::{
+        interactions::allowances::Allowances,
+        liquidity::{
+            uniswap_v3::{self, UniswapV3Liquidity, UniswapV3SettlementHandler},
+            ConcentratedLiquidity,
+        },
+        liquidity_collector::LiquidityCollecting,
+    },
+    std::sync::{Arc, Mutex},
+};
+
+pub fn to_domain(id: liquidity::Id, pool: ConcentratedLiquidity) -> liquidity::Liquidity {
+    let handler = pool
+        .settlement_handling
+        .as_any()
+        .downcast_ref::<uniswap_v3::UniswapV3SettlementHandler>()
+        .expect("downcast uniswap settlement handler");
+
+    liquidity::Liquidity {
+        id,
+        gas: eth::Gas(pool.pool.gas_stats.mean_gas),
+        kind: liquidity::Kind::UniswapV3(Pool {
+            router: handler.inner.router.address().into(),
+            address: pool.pool.address.into(),
+            tokens: liquidity::TokenPair::new(
+                pool.pool.tokens[0].id.into(),
+                pool.pool.tokens[1].id.into(),
+            )
+            .unwrap(),
+            sqrt_price: SqrtPrice(pool.pool.state.sqrt_price),
+            liquidity: Liquidity(pool.pool.state.liquidity.as_u128()),
+            tick: Tick(pool.pool.state.tick.to_i32().unwrap()),
+            liquidity_net: pool
+                .pool
+                .state
+                .liquidity_net
+                .iter()
+                .map(|(key, value)| {
+                    (
+                        Tick(key.to_i32().unwrap()),
+                        LiquidityNet(value.to_i128().unwrap()),
+                    )
+                })
+                .collect(),
+            fee: Fee(pool.pool.state.fee),
+        }),
+    }
+}
+
+pub fn to_interaction(
+    pool: &liquidity::uniswap::v3::Pool,
+    input: &liquidity::MaxInput,
+    output: &liquidity::ExactOutput,
+    receiver: &eth::Address,
+) -> eth::Interaction {
+    let web3 = shared::ethrpc::dummy::web3();
+
+    let handler = UniswapV3SettlementHandler::new(
+        UniswapV3SwapRouter::at(&web3, pool.router.0),
+        GPv2Settlement::at(&web3, receiver.0),
+        Mutex::new(Allowances::empty(receiver.0)),
+        pool.fee.0,
+    );
+
+    let (_, interaction) = handler.settle(
+        TokenAmount::new(input.0.token.into(), input.0.amount),
+        TokenAmount::new(output.0.token.into(), output.0.amount),
+    );
+
+    interaction
+        .encode()
+        .into_iter()
+        .map(|(target, value, call_data)| eth::Interaction {
+            target: eth::Address(target),
+            value: eth::Ether(value),
+            call_data: call_data.0,
+        })
+        .next()
+        .expect("returns exactly 1 interaction")
```

Comment:
If we expect exactly one interaction, we can use [`Itertools::exactly_one`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.exactly_one). As it currently is, I don't think it would care if there is two (so the expectation here is more "returns at least 1 interaction").

Normally, I wouldn't suggest changing this, but since we already pull in `itertools`, I thought it was worth being extra explicit.

Reply:
added: [9bc2433](https://github.com/cowprotocol/services/pull/1251/commits/9bc2433e41bfa34cc029b8f31877af720c6fb959)